gemgis/vector.py

```diff
@@ -19,100 +20,522 @@
+                           overwrite_xy: bool = False,
+                           target_crs: str = None,
+                           bbox: List[Union[int, float]] = None) -> gpd.geodataframe.GeoDataFrame:
+    """
+   Extracting x,y coordinates from a GeoDataFrame (LineStrings) and returning a GeoDataFrame with x,y
+   coordinates as additional columns
+   Args:
+       gdf (gpd.geodataframe.GeoDataFrame): GeoDataFrame created from vector data containing elements of geom_type
+       LineString
+       reset_index (bool): Variable to reset the index of the resulting GeoDataFrame, default True
+       drop_id (bool): Variable to drop the id column, default True
+       drop_index (bool): Variable to drop the index column, default True
+       drop_points (bool): Variable to drop the points column, default True
+       overwrite_xy (bool): Variable to overwrite existing X and Y values, default False
+       target_crs (str, pyproj.crs.crs.CRS): Name of the CRS provided to reproject coordinates of the GeoDataFrame
+       bbox (list): Values (minx, maxx, miny, maxy) to limit the extent of the data
+   Return:
+       gdf (gpd.geodataframe.GeoDataFrame): GeoDataFrame with appended x,y columns and optional columns
+   """
+
+    # Checking that gdf is of type GepDataFrame
+    if not isinstance(gdf, gpd.geodataframe.GeoDataFrame):
+        raise TypeError('Loaded object is not a GeoDataFrame')
+
+    # Check that all entries of the gdf are of type LineString
+    if not all(gdf.geom_type == 'LineString'):
+        raise TypeError('All GeoDataFrame entries must be of geom_type linestrings')
+
+    # Checking that the bbox is of type None or list
+    if not isinstance(bbox, (type(None), list)):
+        raise TypeError('The bbox values must be provided as list')
+
+    # Checking that the bbox list only has four elements
+    if isinstance(bbox, list) and not len(bbox) == 4:
+        raise ValueError('Provide minx, maxx, miny and maxy values for the bbox')
+
+    # Checking that all elements of the list are of type int or float
+    if isinstance(bbox, list) and not all(isinstance(i, (int, float)) for i in bbox):
+        raise TypeError('Bbox values must be of type float or int')
+
+    # Checking that drop_index is of type bool
+    if not isinstance(drop_index, bool):
+        raise TypeError('Drop_index argument must be of type bool')
+
+    # Checking that drop_id is of type bool
+    if not isinstance(drop_id, bool):
+        raise TypeError('Drop_id argument must be of type bool')
+
+    # Checking that drop_points is of type bool
+    if not isinstance(drop_points, bool):
+        raise TypeError('Drop_points argument must be of type bool')
+
+    # Checking that reset_index is of type bool
+    if not isinstance(reset_index, bool):
+        raise TypeError('Reset_index argument must be of type bool')
+
+    # Checking that the target_crs is of type string
+    if not isinstance(target_crs, (str, type(None), pyproj.crs.crs.CRS)):
```

Comment:
```suggestion
    if target_crs is not None and not isinstance(target_crs, (str, pyproj.crs.crs.CRS)):
```

Reply:
As a sidenote, pyproj kinda gave up on annotations for this class and just set use `Any` for the type of projparam:

https://github.com/pyproj4/pyproj/blob/20ffd6f630c46e77286f2550566b2f6aded4cd69/pyproj/crs/crs.py#L278-L294

So they do type checking at runtime to dispatch the correct way to parse the given parameters, and raise CRSError if not found. Maybe we could use that as well: check if it's not an instance of CRS, try to instance a CRS from it and either catch CRSError to make our own error message or just let the error bubble up to the user.

Reply:
```suggestion
    if target_crs is not None:
        if not isinstance(target_crs, pyproj.crs.crs.CRS):
            try:
                target_crs = pyproj.crs.crs.CRS(target_crs)
            except pyproj.exceptions.CRSError:
                raise TypeError('target_crs must either be a pyproj CRS or a valid input for it')
```